cpp/src/arrow/buffer.h

```diff
@@ -259,6 +259,20 @@ class ARROW_EXPORT ResizableBuffer : public MutableBuffer {
   ResizableBuffer(uint8_t* data, int64_t size) : MutableBuffer(data, size) {}
 };
 
+/// \class STLStringBuffer
+/// \brief A Buffer which owns the memory pointed to by a std::string
+class ARROW_EXPORT STLStringBuffer : public Buffer {
+public:
+  STLStringBuffer(std::string& data)
+  : Buffer(reinterpret_cast<const uint8_t*>(std::move(data.c_str())),
+          static_cast<int64_t>(data.size())) {
```

Comment:
This isn't right. This class does not take ownership of the memory in `data`

Reply:
``STLStringBuffer`` will initialize its base class's ``data_`` member with the rvalue reference returned by ``std::move``. Post the operation, ``data_`` member in the class should be the only one with a valid reference to the memory.

Is gaining ownership of the actual data i.e. ``data.c_str()`` the objective, or would we rather have everything that ``data`` points to? I took inspiration from ``FromString`` methods and took reference for only the ``char`` part of the ``string object``

Reply:
`Buffer` won't deallocate this memory, though, because it has a trivial destructor. The safest thing would be to use a `std::string` to transfer the internal state of the input `std::string` and have deallocation handled by RAII